token/jwt/token.go

```diff
@@ -0,0 +1,229 @@
+	}
+}
+
+func (t *Token) toJoseHeader() map[jose.HeaderKey]interface{} {
+	h := map[jose.HeaderKey]interface{}{}
+	for k, v := range t.Header {
+		h[jose.HeaderKey(k)] = v
+	}
+	return h
+}
+
+// SignedString provides a compatible `jwt-go` Token.SignedString method
+//
+// > Get the complete, signed token
+func (t *Token) SignedString(k interface{}) (rawToken string, err error) {
+	if _, ok := k.(unsafeNoneMagicConstant); ok {
+		rawToken, err = unsignedToken(t)
+		return
+
+	}
+	var signer jose.Signer
+	key := jose.SigningKey{
+		Algorithm: t.Method,
+		Key:       k,
+	}
+	opts := &jose.SignerOptions{ExtraHeaders: t.toJoseHeader()}
+	signer, err = jose.NewSigner(key, opts)
+	if err != nil {
+		err = errorsx.WithStack(err)
+		return
+	}
+
+	// A explicit conversion from type alias MapClaims
+	// to map[string]interface{} is required because the
+	// go-jose CompactSerialize() only support explicit maps
+	// as claims or structs but not type aliases from maps.
+	claims := map[string]interface{}(t.Claims)
+	rawToken, err = jwt.Signed(signer).Claims(claims).CompactSerialize()
+	if err != nil {
+		err = &ValidationError{Errors: ValidationErrorClaimsInvalid, Inner: err}
+		return
+	}
+	return
+}
+
+func unsignedToken(t *Token) (string, error) {
+	t.Header["alg"] = "none"
+	hbytes, err := json.Marshal(&t.Header)
+	if err != nil {
+		return "", errorsx.WithStack(err)
+	}
+	bbytes, err := json.Marshal(&t.Claims)
+	if err != nil {
+		return "", errorsx.WithStack(err)
+	}
+	h := base64.RawURLEncoding.EncodeToString(hbytes)
+	b := base64.RawURLEncoding.EncodeToString(bbytes)
+	return fmt.Sprintf("%v.%v.", h, b), nil
+}
+
+func newToken(parsedToken *jwt.JSONWebToken, claims MapClaims) (*Token, error) {
+	token := &Token{Claims: claims}
+	if len(parsedToken.Headers) != 1 {
+		return nil, &ValidationError{text: fmt.Sprintf("only one header supported, got %v", len(parsedToken.Headers)), Errors: ValidationErrorMalformed}
+	}
+
+	// copy headers
+	h := parsedToken.Headers[0]
+	token.Header = map[string]interface{}{
+		"alg": h.Algorithm,
+	}
+	if h.KeyID != "" {
+		token.Header["kid"] = h.KeyID
+	}
+	for k, v := range h.ExtraHeaders {
+		token.Header[string(k)] = v
+	}
+
+	token.Method = jose.SignatureAlgorithm(h.Algorithm)
+
+	return token, nil
+}
+
+// Parse methods use this callback function to supply
+// the key for verification.  The function receives the parsed,
+// but unverified Token.  This allows you to use properties in the
+// Header of the token (such as `kid`) to identify which key to use.
+type Keyfunc func(*Token) (interface{}, error)
+
+func Parse(tokenString string, keyFunc Keyfunc) (*Token, error) {
+	return ParseWithClaims(tokenString, MapClaims{}, keyFunc)
+}
+
+// Parse, validate, and return a token.
+// keyFunc will receive the parsed token and should return the key for validating.
+// If everything is kosher, err will be nil
+func ParseWithClaims(rawToken string, claims MapClaims, keyFunc Keyfunc) (*Token, error) {
```

Comment:
I think we need tests for this, as we're using it to verify authentication tokens and it looks like it has several moving parts. I think as part of the tests we should have:

1. Completely invalid token like`asdf`
2. Token that is not a JWT `foo.bar.baz`
3. Token that mismatches the keyfunc:
   1. Expect RSA256 but get RSA512
   2. Expect RSA256 but get None
5.  RSA256 Token with unknown signing key
6. Token with invalid claims:
   1. Expired
   1. Incorrect IAT
   2. ...and other cases that should be covered by this

And finally also some passing tokens like:

1. RSA
2. None
3. ECDSA

Reply:
done